File: LeetCode/17. Letter Combinations of a Phone Number/Solution.java
class Solution {
    String[] telMap = {"_","!@#","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
    List<String> res = new ArrayList<String>();
    public List<String> letterCombinations(String digits) {
        if(digits == null || digits.length() == 0){
            return res;
        }
        process(digits, "", 0);

        return res;
    }

    private void process(String digits, String letter, int index){
          // 递归的第一步，确定终止条件
          if(index == digits.length()){
            res.add(letter);
            return;
        }
        String str = telMap[digits.charAt(index)-'0'];
        for(int i=0; i<str.length();i++){
            process(digits, letter+str.charAt(i), index+1);
        }
    }
}

File: LeetCode/17. Letter Combinations of a Phone Number/Solution.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        unordered_map<char, string> table{
            {'0', " "}, {'1',"*"}, {'2', "abc"},
            {'3',"def"}, {'4',"ghi"}, {'5',"jkl"},
            {'6',"mno"}, {'7',"pqrs"},{'8',"tuv"},
            {'9',"wxyz"}};  
        vector<string> res;
        if(digits == "") return res;
        func(res, "", digits, table, 0);
        return res;
    }
    
    void func(vector<string> &res, string str, string &digits, unordered_map<char, string> &m, int index){
        if(str.size() == digits.size()){
            res.push_back(str);
            return;
        }

        for(char c : m[digits[index]]){
            func(res, str+c, digits, m, index+1);
        }
        return;
    }
};